rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+
+2. The second drawback is related to the fact that structs take up more storage slots than they actually need. This behavior is **not** new. There is a tradeoff here to be considered. If we make structs tightly packed, then manipulating a single field becomes more expensive (potential read-modify-write patterns or fields "spilling" over to the next slot), but the number of storage slots required may become lower. We might want to consider a `#[packed]` attribute in the future to let the user decide on one way v.s. the other. Another thing to consider with tightly-packed structs is that the type `StorageHandler` would then also require an additional field that points to a particular location _inside_ a given slot (i.e. which byte to start reading from or writing to).
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+One alternative is to continue with the current use model which has many flaws so that's not ideal. Another alternative is to introduce an `AsStorageHandler` trait that allows calling `storage.x.as_storage_handler()` to get the storage handler manually, but that approach has its limitations when implementing `StorageMap` and `StorageVec` and requires defining the behavior of `as_storage_hander` when called on a regular stack variable.
+
+One more alternative is to completely remove the concept of a "storage variable" and switch over to thinking about storage like file I/O or databases. However, I think the approach described in this RFC provides a good middle ground where methods like `store` and `get` somewhat behave like I/O methods but also keeps the abstraction of "storage variables" to help user write readable code.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+Solidity's approach to storage is completely different from the above and has its own problems, as storage variable are hard to identify and storage accesses are not explicit enough. Sway's current approach is also flawed as explained in the [Motivation](#motivation) section.
+
+In Rust, the closest thing to storage accesses is file I/O where a handler is used to read and write to a file similarly to `store` and `get` above.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- How do we make storage accesses, particularly reads, more ergonomic?
+- Is not having structs tightly packed acceptable? Is that something that we want in the future, potentially via a `#[packed]` attribute that also reorders the fields of the struct for optimal layout?
```

Comment:
This is a good point - most users will just use whatever the default is so it would be good to get it right.

I guess if we're to pack by default, users could always add their own padding to structs with `[u8; N]` if they want to avoid the awkwardness of typed offsets (mentioned above)?

On the other hand, if we want to track offsets at compile-time, we probably can't add this retroactively without breaking types in existing code, so we'd need to add some field to the handle as you mention. As long as the `StorageHandler` type is opaque, maybe that's fine though?

Reply:
Packed structs are now the default.